File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res=[]
        nums.sort()
        for i, val in enumerate(nums):
            if (i>0 and val==nums[i-1]):
                continue
            l=i+1
            r=len(nums)-1
            while l<r:
                summ=val+nums[l]+nums[r]
                if(summ>0):
                    r-=1
                elif (summ<0):
                    l+=1
                else:
                    res.append([val,nums[l],nums[r]])
                    l+=1
                    while(l<r and nums[l]==nums[l-1]):
                        l+=1
        return res

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res=new ArrayList<>();
        Arrays.sort(nums);
        
        for(int i=0;i<nums.length-2;i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }
            int l=i+1;
            int r=nums.length-1;
            while(l<r){
                int summ=nums[i]+nums[l]+nums[r];
                if(summ>0){
                    r--;
                }
                else if(summ<0){
                    l++;
                }
                else{
                    res.add(Arrays.asList(nums[i],nums[l],nums[r]));
                    l++;
                    while(l<r && nums[l]==nums[l-1]){
                        l++;
                    }
                }
            }
        }
        return res;
    }
}